### PR TITLE
Fix build on SLE12

### PR DESCRIPTION
**Pull request by hramrach**

### CODE REVIEW

File: osc/credentials.py [hunk 1]
```diff
@@ -25,7 +27,6 @@
     print(msg, e, file=sys.stderr)
     gnomekeyring = None
 
-from . import conf
```
Comment (dmach) [inline osc/credentials.py:28]: The `conf` module is used later in the code, removing the line leads to the following errors found by pylint:
```
osc/credentials.py:234:42: E0602: Undefined variable 'conf' (undefined-variable)
osc/credentials.py:420:38: E0602: Undefined variable 'conf' (undefined-variable)
```

**Review by dmach — CHANGES_REQUESTED** [review]

**Comment by hramrach** [comment]
Nonetheless it fails to load on SLE12:

[   30s] + python3 setup.py build
[   30s] Traceback (most recent call last):
[   30s]   File "setup.py", line 11, in <module>
[   30s]     import osc.core
[   30s]   File "/home/abuild/rpmbuild/BUILD/osc-0.181.0/osc/core.py", line 62, in <module>
[   30s]     from . import conf
[   30s]   File "/home/abuild/rpmbuild/BUILD/osc-0.181.0/osc/conf.py", line 73, in <module>
[   30s]     from osc import credentials
[   30s]   File "/home/abuild/rpmbuild/BUILD/osc-0.181.0/osc/credentials.py", line 32, in <module>
[   30s]     from . import conf
[   30s] ImportError: cannot import name 'conf'
[   30s] error: Bad exit status from /var/tmp/rpm-tmp.j407Zp (%build)

https://build.opensuse.org/package/live_build_log/home:michals/osc/SLE_12_SP5/x86_64

**Comment by dmach** [comment]
I couldn't update your PR because you've submitted it from master,
so I'm pasting the changes that seem to work for me here.
It's also fixing #1123
I've created `0.x-maint` branch this PR should go to, because we probably no longer want such changes in master that has a completely different code base now.

```
diff --git a/.github/workflows/unittests.yaml b/.github/workflows/unittests.yaml
index 26195bc7..e8256108 100644
--- a/.github/workflows/unittests.yaml
+++ b/.github/workflows/unittests.yaml
@@ -50,7 +50,7 @@ jobs:
             zypper --non-interactive --gpg-auto-import-keys refresh
             zypper --non-interactive dist-upgrade
             zypper --non-interactive install git-lfs
-            zypper --non-interactive install diffstat diffutils python3 python3-chardet python3-M2Crypto python3-pip python3-rpm python3-setuptools
+            zypper --non-interactive install diffstat diffutils python3 python3-M2Crypto python3-pip python3-rpm python3-setuptools
 
       - name: 'Install packages (Fedora/CentOS)'
         if: ${{ startsWith(matrix.container, 'fedora:') || contains(matrix.container, 'centos:')  }}
@@ -58,7 +58,7 @@ jobs:
             dnf -y makecache
             dnf -y distro-sync
             dnf -y install git-lfs
-            dnf -y install diffstat diffutils python3 python3-chardet python3-m2crypto python3-pip python3-rpm python3-setuptools
+            dnf -y install diffstat diffutils python3 python3-m2crypto python3-pip python3-rpm python3-setuptools
 
       - name: 'Install packages (Debian/Ubuntu)'
         if: ${{ startsWith(matrix.container, 'debian:') || startsWith(matrix.container, 'ubuntu:')  }}
@@ -66,7 +66,7 @@ jobs:
             apt-get -y update
             apt-get -y upgrade
             apt-get -y --no-install-recommends install git-lfs
-            apt-get -y --no-install-recommends install diffstat diffutils python3 python3-chardet python3-m2crypto python3-pip python3-rpm python3-setuptools
+            apt-get -y --no-install-recommends install diffstat diffutils python3 python3-m2crypto python3-pip python3-rpm python3-setuptools
 
       - uses: actions/checkout@v3
 
diff --git a/osc/conf.py b/osc/conf.py
index 6d8cc1cb..e3be9297 100644
--- a/osc/conf.py
+++ b/osc/conf.py
@@ -537,7 +537,11 @@ def _build_opener(apiurl):
         def http_error_401(self, req, fp, code, msg, headers):
             self._rewind_request(req)
             authreqs = {}
-            for authreq in headers.get_all('www-authenticate', []):
+            if hasattr(headers, "get_all"):
+                all_headers = headers.get_all('www-authenticate', [])
+            else:
+                all_headers = headers.getallmatchingheaders('www-authenticate')
+            for authreq in all_headers:
                 scheme = authreq.split()[0].lower()
                 authreqs[scheme] = authreq
 
diff --git a/osc/credentials.py b/osc/credentials.py
index d076b2ad..30242b6d 100644
--- a/osc/credentials.py
+++ b/osc/credentials.py
@@ -1,3 +1,5 @@
+from __future__ import print_function
+
 import importlib
 import bz2
 import base64
@@ -29,7 +31,6 @@ except BaseException as e:
     print(msg, e, file=sys.stderr)
     gnomekeyring = None
 
-from . import conf
 from . import oscerr
 
 
@@ -234,6 +235,7 @@ class KeyringCredentialsManager(AbstractCredentialsManager):
             keyring_backend = keyring.core.load_keyring(self._backend_cls_name)
         except ModuleNotFoundError:
             msg = "Invalid credentials_mgr_class: {}".format(self._backend_cls_name)
+            from . import conf
             raise oscerr.ConfigError(msg, conf.config['conffile'])
         keyring.set_keyring(keyring_backend)
 
@@ -420,6 +422,7 @@ def create_credentials_manager(url, cp):
         creds_mgr = getattr(importlib.import_module(mod), cls).create(cp, options)
     except ModuleNotFoundError:
         msg = "Invalid credentials_mgr_class: {}".format(creds_mgr_cls)
+        from . import conf
         raise oscerr.ConfigError(msg, conf.config['conffile'])
     return creds_mgr
 
diff --git a/setup.py b/setup.py
index e0dd9b76..94c6596e 100755
--- a/setup.py
+++ b/setup.py
@@ -103,7 +103,7 @@ setuptools.setup(
     packages=['osc', 'osc.util'],
     scripts=['osc-wrapper.py'],
     data_files=data_files,
-    install_requires=['M2Crypto', 'chardet'],
+    install_requires=['M2Crypto'],
     classifiers=[
         "Development Status :: 5 - Production/Stable",
         "Environment :: Console",
@@ -129,5 +129,4 @@ setuptools.setup(
     ],
     # Override certain command classes with our own ones
     cmdclass=cmdclass,
-    test_suite="tests",
 )

```

**Comment by hramrach** [comment]
Thanks for the patch. I can run the tests not but I still see errors:

======================================================================
ERROR: testDiffReplacedEmptyTodo (tests.test_difffiles.TestDiffFiles)
diff a complete package
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/suse/msuchanek/openSUSE.org:Kernel:tools/osc/osc-0.181.0/tests/test_difffiles.py", line 157, in testDiffReplacedEmptyTodo
    self.__check_diff(p, exp, None)
  File "/suse/msuchanek/openSUSE.org:Kernel:tools/osc/osc-0.181.0/tests/test_difffiles.py", line 301, in __check_diff
    for i in p.get_diff(revision):
  File "/suse/msuchanek/openSUSE.org:Kernel:tools/osc/osc-0.181.0/osc/core.py", line 2111, in get_diff
    yield [diff_hdr % f.name.encode()]
TypeError: unsupported operand type(s) for %: 'bytes' and 'bytes'

**Comment by dmach** [comment]
> Thanks for the patch. I can run the tests not but I still see errors:

```
...
    yield [diff_hdr % f.name.encode()]
TypeError: unsupported operand type(s) for %: 'bytes' and 'bytes'
```

That's most likely something related to `python3`.
Tests pass on SLE 12 when I use `python`. Could you try `python`?

**Comment by hramrach** [comment]
Yes, it works with python

**Comment by hramrach** [comment]
The fix for o.181 is here https://github.com/openSUSE/osc/pull/1124

It might be possible to forward-port to current code but this PR is not complete

**Pull request closed without merging**